### ISSUE TITLE
sep-by / any-string combo doesn't play nicely with EOF

### ISSUE THREAD

**Issue by mccraigmccraig** [protoflex/parse-ez#5]
i couldn't make the sep-by/any-string combination in the example csv parser work nicely when there isn't a record terminator at  EOF :

(use 'protoflex.examples.csv_parse)

(parse csv "foo,bar\nbaz,boo\n")
;; [["foo" "bar"] ["baz" "boo"]]

(parse csv "foo,bar\nbaz,boo")
Exception Parse Error: Extraneous text at line 2, col 1  protoflex.parse/throw-ex (parse.clj:274)

**Comment by protoflex** [comment]
Yes, any-string expects a seperator or EOL to delimit plain (unquoted) strings.
Will need to either change the behavior of any-string or add another version of any-string that will accept EOF as valid delimiter.

Thanks.

**Comment by protoflex** [comment]
Will add a fix to next release. Until then, here is the patch: (1 new fn "read-to-re-or-eof", and two changed functions: "any-string" and "sep-by"):

``` clojure
(defn read-to-re-or-eof
  "If the specified regex matches in the remaining text, returns text upto the match;
   Otherwise returns all the remaining text and the input cursor is positioned at EOF."
  [re]
  (if-let [t (attempt #(read-to-re re))] t (to-eof)))

(defn any-string
  "Reads a single-quoted or double-quoted or a plain-string that is followed
  by the specified separator sep or EOF; the separator is not part of the returned
  string."
  [sep] (cond
          (starts-with? "\"") (dq-str)
          (starts-with? "'") (sq-str)
          :else (read-to-re-or-eof (re-pattern (str sep "|\r?\n")))))

(defn sep-by
  "Reads a record using the specified field, field-separator and 
  record-separator parse functions.  If no record-separator is specified, 
  a newline character is used as record separator. Returns the fields of the 
  record in a vector."
  ([fld-fn fld-sep-fn]
   (sep-by fld-fn fld-sep-fn #(regex #"\r?\n")))

  ([fld-fn fld-sep-fn rec-sep-fn]
    (if-not (at-end?)
      (let [fst (fld-fn)
            rst (multi* #(series fld-sep-fn fld-fn))
            _ (any rec-sep-fn at-end?)
            rst (if rst (map second rst))
            result (if rst (vec(conj rst fst)) [fst])]
        result))))
```

**Comment by mccraigmccraig** [comment]
ah, i see : i had a go at a solution, but i failed to get a working function like read-to-re-or-eof : thanks !

**Comment by protoflex** [comment]
Also, the sep-by will keep matching empty input (when EOF is accepted as a delimiter), so I added an "if-not (at-end?)" check in sep-by.

**Comment by protoflex** [comment]
Fixed in 0.4.1.
